### PR TITLE
Create sip-alg plugin documentation

### PR DESCRIPTION
**Pull request by agrawalkaushik**
DOCS-110 #time 1d

### CODE REVIEW

File: docs/plugin_sip_alg.md [hunk 1]
```diff
@@ -0,0 +1,473 @@
...
+
+### Inbound and Outbound Services
+
+The `sip-alg` plugin operates on the concept of [SFC](plugin_intro.md#service-function-chaining) where all traffic is routed using KNIs to Linux. The 128T router uses `kamailio`[^3] to then [fix all SIP messages](#sip-message-changes) and perform additional routing to manage both inbound and outbound calls. For each of the `inbound` and `outbound` call the plugin requires two services - an `ingress service` to route the traffic to the KNI and an `egress service` to route the traffic to the other SIP endpoint. The user is responsible for configuring the `egress service` and corresponding routes. The plugin automatically generates the `ingress service` and associated routes using the egress service and plugin configuration.
```
Comment (MichaelBaj) [inline docs/plugin_sip_alg.md:94]: .. where all traffic is routed using KNIs to and from Linux.
For each of the `inbound` and `outbound` calls the plugin requires two services [[missing plurality]]

File: docs/plugin_sip_alg.md [hunk 2]
```diff
@@ -0,0 +1,473 @@
...
+
+#### Outbound Service and Config Auto Generation
+
+The `outbound` service is used for initial registration between the branch PBX and the SIP server as well as making all outbound calls to remote endpoints. A typical flow for such calls start from the private lan network, routed via the 128T sip-alg function and the fixed up SIP messages are then routed over the wan interface to the remote endpoint. The user is responsible for configuring the egress service and the associated routing policies such as SVR etc. Consider the following example:
```
Comment (MichaelBaj) [inline docs/plugin_sip_alg.md:98]: A typical call flow originates from the private LAN network, routed via the 128T sip-alg function that transforms the SIP messages, which are then routed over the WAN interface to the remote endpoint.

File: docs/plugin_sip_alg.md [hunk 3]
```diff
@@ -0,0 +1,473 @@
...
+
+        access-policy  _internal_
+            source      _internal_
+            permission  allow
+        exit
+    exit
+exit
+```
+
+The service `sip-outbound-dc1` captures all SIP udp traffic towards port 5060 originating from the `_internal_` tenant. This is a special tenant associated with the SFC KNIs. The `sip-service > outbound` for the sip-alg configuration refers to this user configured service in the above example. In this process, the plugin will inherit all the address and transport configuration from this service and combined with the defined access-policy on the plugin will generate the configuration for SIP traffic received on the lan tenant.
```
Comment (MichaelBaj) [inline docs/plugin_sip_alg.md:126]: UDP should be capitalized.

```[`_internal_` tenant](bcp_tenants.mdx#the-internal-tenant)```

... and transport configuration from this _service_, combined with the defined _access-policy_ on the plugin, will generate the configuration for the SIP traffic received on the LAN tenant.

File: docs/plugin_sip_alg.md [hunk 4]
```diff
@@ -0,0 +1,473 @@
...
+                                                                                                                                                                                                                                             0:00:16
+
+Completed in 0.04 seconds
+admin@node1.router1#
+
+```
+
+#### Inbound Service and Config Auto Generation
+
+The `inbound` service is used for processing all incoming calls to the branch PBX from the remote server. A typical flow for such calls start from the public wan network, routed via the 128T sip-alg function and the fixed up SIP messages are then routed over the lan interface to the branch PBX. The user is responsible for configuring the ingress service and the associated service-routes. Consider the following example:
```
Comment (MichaelBaj) [inline docs/plugin_sip_alg.md:166]: A typical call flow originates from the public WAN network, routed via the 128T sip-alg function that transforms the SIP messages, which are then routed over the LAN interface to the branch PBX.

File: docs/plugin_sip_alg.md [hunk 5]
```diff
@@ -0,0 +1,473 @@
...
+
+        access-policy  _internal_
+            source      _internal_
+            permission  allow
+        exit
+    exit
+exit
+```
+
+The service `sip-inbound-pbx` captures all SIP udp traffic towards port 5060 originating from the `_internal_` tenant. This is a special tenant associated with the SFC KNIs. The `sip-service > inbound` for the sip-alg configuration refers to this user configured service in the above example. In this process, the plugin will inherit all the address and transport configuration from this service and combined with the defined access-policy on the plugin will generate the configuration for SIP traffic received on the `datacenter` tenant.
```
Comment (MichaelBaj) [inline docs/plugin_sip_alg.md:195]: UDP should be capitalized

```[`_internal_` tenant](bcp_tenants.mdx#the-internal-tenant)```

... and transport configuration from this service, combined with the defined access-policy on the plugin, will generate the configuration for the SIP traffic received on the `datacenter` tenant.

File: docs/plugin_sip_alg.md [hunk 6]
```diff
@@ -0,0 +1,473 @@
...
+ f62248d7-a003-4e6b-b586-d196c4b23d3d   rev   sip-inbound-pbx                                    _internal_   lan               0   udp     192.168.2.128       5060   172.16.2.128        5060   0.0.0.0          0   false          3538   0 days
+                                                                                                                                                                                                                                            0:01:07
+
+Completed in 0.05 seconds
+admin@node1.router1#
+```
+
+### Remove SDP Prefixes
+
+Some of the SDP options can include private IP addresses and pose additional challenges with the SIP NAT inter-working. The `remove-sdp-lines-by-prefix` can be used to remove such lines from the SIP messages. As shown in the [above example](#configuration), the `a=candidate` line is configured to be removed from all SIP messages.
```
Comment (MichaelBaj) [inline docs/plugin_sip_alg.md:234]: Is it worthwhile to enumerate all SDP options in this document, or provide a link to the list in the Kamilio documentation?

File: docs/plugin_sip_alg.md [hunk 7]
```diff
@@ -0,0 +1,473 @@
...
+
+## SIP Message Changes
+
+The 128T `sip-alg` functionality is completely stateless and works by applying the NAT rules to every SIP message sent and received. The SIP stack, however, is aware of directionality and uses this information to fix the messages with the correct set of IP addresses.
```
Comment (MichaelBaj) [inline docs/plugin_sip_alg.md:238]: ... and works by applying NAT rules to every SIP message ...

File: docs/plugin_sip_alg.md [hunk 8]
```diff
@@ -0,0 +1,473 @@
...
+Jul 08 18:57:36 t190-dut2.openstacklocal PBX[22278]: INFO: {2 Cseq: 2 BYE CId: 5-12741@192.168.2.128} <script>: [ STATUS 200 OK ]
+Jul 08 18:57:36 t190-dut2.openstacklocal PBX[22278]: INFO: {2 Cseq: 2 BYE CId: 5-12741@192.168.2.128} <script>: [ VIA SIP/2.0/UDP 172.16.2.128:5061;branch=z9hG4bK4911.716b96ecc6684d3543e468d8454f285e.0, SIP/2.0/UDP 172.16.1.1...hG4bK-12741-5-7 ]
+Jul 08 18:57:36 t190-dut2.openstacklocal PBX[22278]: INFO: {2 Cseq: 2 BYE CId: 5-12741@192.168.2.128} <script>: [ FROM sipp <sip:sipp@192.168.2.128:5060>;tag=12741SIPpTag005 ]
+Jul 08 18:57:36 t190-dut2.openstacklocal PBX[22278]: INFO: {2 Cseq: 2 BYE CId: 5-12741@192.168.2.128} <script>: [ TO service <sip:service@172.16.2.102:5061>;tag=11336SIPpTag015 ]
+Jul 08 18:57:36 t190-dut2.openstacklocal PBX[22278]: INFO: {2 Cseq: 2 BYE CId: 5-12741@192.168.2.128} <script>: [ CONTACT <sip:172.16.2.102:5061;transport=UDP> ]
+Hint: Some lines were ellipsized, use -l to show in full.
+#
+```
+
+The `kamailio` configuration can be found under `/etc/kamailio/kamailio.PBX.cfg` and should reflect the ip-mappings and other configuration from the plugin configuration.
```
Comment (MichaelBaj) [inline docs/plugin_sip_alg.md:406]: should probably add a note not to make changes to this configuration as they will get overwritten by the 128T

File: docs/plugin_sip_alg.md [hunk 9]
```diff
@@ -0,0 +1,473 @@
...
+
+
+### Inspecting the sip-alg Namespace
+
+All of the `sip-alg` function including `kamailio` operate from within a network namespace. Some of the common commands and the expected outputs are shown below
```
Comment (MichaelBaj) [inline docs/plugin_sip_alg.md:411]: All of the `sip-alg` functions, including `kamilio`, operate from ...

Missing period at end of sentence.

File: docs/plugin_sip_alg.md [hunk 10]
```diff
@@ -0,0 +1,473 @@
...
+The `kamailio` process would be listening on the configured SIP port to intercept and fix the SIP messages. The behavior can be validated by running the following command:
+```
+# ip netns exec PBX netstat -pant
+Active Internet connections (servers and established)
+Proto Recv-Q Send-Q Local Address           Foreign Address         State       PID/Program name
+tcp        0      0 169.254.129.22:5060     0.0.0.0:*               LISTEN      19098/kamailio
+tcp        0      0 169.254.129.18:5060     0.0.0.0:*               LISTEN      19098/kamailio
+```
+
+Finally, the namespace is configured with a set of `iptables` rules to enable the static mapping from public to private. A typical set of rules look like below:
```
Comment (MichaelBaj) [inline docs/plugin_sip_alg.md:452]: A typical set of rules can be seen below.

File: docs/plugin_sip_alg.md [hunk 11]
```diff
@@ -0,0 +1,473 @@
...
+
+## Configuration
+The sip-alg configuration can be enabled on a router. In order to configure `sip-alg`, it is important to consider the following:
+
+* Static IP mappings
+* Services for making inbound and outbound calls
+* Additional changes to the SDP to enable SIP inter working
+
+:::important
+The 128T sip-alg function currently only supports a scenario whereby all the SIP endpoints on the private network communicate via a SIP enabled PBX device.
```
[the remark this answers is not in the document]
Comment (agrawalkaushik) [reply docs/plugin_sip_alg.md:23]: Updated the pictures for inbound and outbound call flows to illustrate the point.

File: docs/plugin_sip_alg.md (hunk 6 above)
Comment (agrawalkaushik) [reply docs/plugin_sip_alg.md:234]: Kamailio provides a lot of SDP related functionality but the only one we expose is the stripping SDP lines that match prefix. The `Kamailio` documentation on this is also very "technical" - https://kamailio.org/docs/modules/4.3.x/modules/sdpops.html#sdpops.f.remove_line_by_prefix. I don't think we should include but I'll defer to your judgement.

File: docs/plugin_sip_alg.md (hunk 8 above)
Comment (agrawalkaushik) [reply docs/plugin_sip_alg.md:406]: Good point - done.

File: docs/plugin_sip_alg.md (hunk 6 above)
Comment (agrawalkaushik) [reply docs/plugin_sip_alg.md:234]: @mendezcirera can probably also shed some more light here.

File: docs/plugin_sip_alg.md [hunk 12]
```diff
@@ -0,0 +1,473 @@
...
+
+                egress-service  sip-inbound-pbx
+                    name           sip-inbound-pbx
+
+                    access-policy  datacenter
+                        source      datacenter
+                        permission  allow
+                    exit
+
+                    additional-prefix   192.168.2.128
```
Comment (BenMatase) [inline docs/plugin_sip_alg.md:74]: This config option is `additional-address` even though it is a prefix, it is trying to mimic the `address` field for a `service`.

Comment (BenMatase) [inline docs/plugin_sip_alg.md:74]: Add a section/more detail describing `additional-address` and `ingress-service-policy`?

File: docs/bcp_tenants.mdx [hunk 13]
```diff
@@ -303,7 +303,7 @@ Node: node1
 
 #### The Internal Tenant
 
-The internal tenant (named `_internet_`), is arguably the most useful of all of the special tenants. It is immutably assigned to the built-in Kernel NIC Interface (KNI) named `kni254`, which is used for communication to and from the Linux host operating system on the platform on which the 128T software is running.
+The internal tenant (named `_internal_`), is arguably the most useful of all of the special tenants. It is immutably assigned to the built-in Kernel NIC Interface (KNI) named `kni254`, which is used for communication to and from the Linux host operating system on the platform on which the 128T software is running.
```
Comment (MichaelBaj) [inline docs/bcp_tenants.mdx:306]: provide a link to `_internal_`
```
[`internal`](bcp_tenants.mdx#the-internal-tenant)
```

File: docs/plugin_sip_alg.md [hunk 14]
```diff
@@ -0,0 +1,473 @@
...
+
+Chain OUTPUT (policy ACCEPT 10 packets, 5485 bytes)
+ pkts bytes target     prot opt in     out     source               destination
+
+Chain POSTROUTING (policy ACCEPT 0 packets, 0 bytes)
+ pkts bytes target     prot opt in     out     source               destination
+   10  5485 SNAT       all  --  *      *       0.0.0.0/0            0.0.0.0/0            to:172.16.2.128
+```
+
+[^1]: https://en.wikipedia.org/wiki/Session_Initiation_Protocol
```
Comment (BenMatase) [inline docs/plugin_sip_alg.md:475]: This works, but I do like the named links like we've used in other MD docs

File: docs/plugin_sip_alg.md (hunk 12 above)
Comment (agrawalkaushik) [reply docs/plugin_sip_alg.md:74]: I'll update to show the correct value.

Comment (agrawalkaushik) [reply docs/plugin_sip_alg.md:74]: I mention how to use `additional-address` in one of the foot notes. I can add a blurb about service policy though.

Comment (BenMatase) [reply docs/plugin_sip_alg.md:74]: Ahh, I think I missed it because it is `additional-prefix`. Yea, I think a blurb about service policy would be good.

**Pull request merged**